### PR TITLE
instant track export

### PR DESCRIPTION
**Pull request by thrillfall**
**Describe the pull request**
Instant post-workout track export to external storage

**Link to the the issue**
#186 #90

### CODE REVIEW

File: src/main/java/de/dennisguse/opentracks/io/file/TrackFileFormat.java [hunk 1]
```diff
@@ -22,7 +22,7 @@
     KML_ONLY_TRACK {
         @Override
         public TrackWriter newTrackWriter(Context context, boolean multiple) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/TrackFileFormat.java]: Then `boolean multiple` is not required a parameter, right?

PS: Please format files (simples solution AndroidStudio can do this automatically before commit).

File: src/androidTest/java/de/dennisguse/opentracks/io/file/importer/ExportImportTest.java [hunk 2]
```diff
@@ -234,7 +234,7 @@ public void gpx() {
         // when
         // 1. export
         ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-        trackExporter.writeTrack(outputStream);
+        trackExporter.writeTrack(new Track[]{track}, outputStream);
```
Comment (dennisguse) [inline src/androidTest/java/de/dennisguse/opentracks/io/file/importer/ExportImportTest.java:237]: Why not overload writeTrack() directly in TrackExporter?
* writeTrack(Track track, OutputStream outputStream);
* writeTrack(Track[] tracks, OutputStream outputStream);

File: src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent resultData) {
+        super.onActivityResult(requestCode, resultCode, resultData);
+        if (requestCode == DIRECTORY_PICKER_REQUEST_CODE) {
+            if (resultCode == Activity.RESULT_OK) {
+                Uri directoryUri = resultData.getData();
+
+                SharedPreferences sharedPreferences = PreferencesUtils.getSharedPreferences(getApplicationContext());
+                SharedPreferences.Editor editor = sharedPreferences.edit();
+                editor.putString(getString(R.string.settings_single_export_directory_key), directoryUri.toString());
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java:38]: Can you move this as a separate method into PreferencesUtils?
It is kind of messy, but at least all preference access is in one file.

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        this.contentProviderUtils = contentProviderUtils;
+        this.sharedPreferences = sharedPreferences;
+        this.trackFileFormat = trackFileFormat;
+    }
+
+    public void exportLastTrackToExternalStorage(String uriAsString) {
+        String directoryAsString = sharedPreferences.getString(uriAsString, "invalid uri");
+        Uri directoryUri = Uri.parse(directoryAsString);
+        DocumentFile directory = DocumentFile.fromTreeUri(context, directoryUri);
+        assert directory != null;
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java:40]: Not necessary as we would die in line 45 with an NPE anyhow:
```
         DocumentFile file = directory.findFile(fileName);
```

File: src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java [hunk 5]
```diff
@@ -164,17 +166,26 @@ public void onDestroy() {
...
             voiceExecutor.shutdown();
         } finally {
             voiceExecutor = null;
         }
 
+        if (PreferencesUtils.shouldInstantExportAfterWorkout(getApplicationContext())) {
+            SharedPreferences sharedPreferences = PreferencesUtils.getSharedPreferences(this);
+            TrackFileFormat trackFileFormat = PreferencesUtils.getExportTrackFileFormat(this);
+            InstantPostWorkoutExport instantPostWorkoutExport = new InstantPostWorkoutExport(getApplicationContext(), contentProviderUtils, sharedPreferences, trackFileFormat);
+            instantPostWorkoutExport.exportLastTrackToExternalStorage(getString(R.string.settings_single_export_directory_key));
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java:180]: use PreferencesUtils

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import de.dennisguse.opentracks.R;
+import de.dennisguse.opentracks.content.provider.ContentProviderUtils;
+import de.dennisguse.opentracks.io.file.TrackFileFormat;
+import de.dennisguse.opentracks.util.StringUtils;
+
+public class InstantPostWorkoutExport {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java:20]: Can we not reuse ExportActivity?

**Comment by dennisguse** [comment]
@thrillfall code looks pretty good.
Can you split this PR into two:
1. refactoring of the internals, and
2. automatic export on stop.

The refactoring looks pretty straight forward and some tiny improvements are required before merging (really, it is just nitpicking).
About the automatic export: I would prefer to re-use ExportActivity if possible; if not possible, then at least the same code should be used (e.g., use same file names etc).
I guess using the introduced export folder setting would even allow us to not ask the user every time, when he/she/* triggers export all from the settings.

PS: Good job!

File: src/androidTest/java/de/dennisguse/opentracks/io/file/importer/ExportImportTest.java (hunk 2 above)
Comment (thrillfall) [reply src/androidTest/java/de/dennisguse/opentracks/io/file/importer/ExportImportTest.java:237]: it seemed wrong to me to have both exporters implement the very same method but i also did not want to create an abstract class so i just left it as is.
i now added the overload

File: src/main/java/de/dennisguse/opentracks/io/file/TrackFileFormat.java (hunk 1 above)
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/TrackFileFormat.java]: right

**Comment by thrillfall** [comment]
@dennisguse thanks for the review. I created a new MR containing only the refactoring. Once this has been merged. I'll create a MR for the instant exporter.

File: src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java (hunk 3 above)
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java:38]: found a method there that does exactly the thing

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java (hunk 4 above)
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java:40]: android studio suggested this. i removed it now

File: src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java (hunk 5 above)
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java:180]: what do you mean by this?

Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java:180]: Do not pass the sharedPreferences to InstantPostWorkoutExport to retrieve infos.
Just put helper functions into PreferenceUtils that get you the data you need.

**Pull request closed without merging**